Better handling of std values and fix bug in pull request # 48

signo opened the pull request:
Reverts pull request #47, and implements a better handling of std values by adding options std_show and std_mark in order to let developer decide if he wants to show the std values or not and also put a mark to visually indicate fields using std values.

This commit also fixes a bug in pull request # 48: the filter parameters where in the wrong order, since the first parameter must be the default value to return.

Commit:
Reverts pull request #47, and implements a better handling of std values by adding options std_show and std_mark in order to let developer decide if he wants to show the std values or not and also put a mark to visually indicate fields using std values.

This commit also fixes a bug in pull request # 48: the filter parameters where in the wrong order, since the first parameter must be the default value to return.

denis0706 commented:
Hi.
Mark isn't put, cause `$std_mark` is always set to empty string.
It happens because  `function _register_setting()` fires after  `function _set_default_options()`, 
so `isset($this->options[$field['id']])`  returns true at `function _register_setting()`

Also at `function _field_input()`, if `std_show` set to true, chechbox fields with std=1 won't be able to set 0.

signo commented:
Hi Denis, std_mark is empty on purpose, in order to preserve the way the plugin is behaving on the installed base, to turn on showing the std value you have to explicitly set std_show to true and for those that want to show the mark they will have to set std mark to a non empty value, i prefer " (*)". If you guys feel the old behavior was wrong we could set std_on and set a value to std_mark by default in 2.0.

About checkbox fields, the way php compares values if we use == both true and 1 will be true, it is only if we use === (strict compare) that a value of 1 will be ignored.

Cheers.

denis0706 commented:
It happens when you set std_show to true and std_mark is not empty.
`$std_mark = ( !isset($this->options[$field['id']]) && $this->args['std_show'] == true && isset($field['std']) ) 
? $this->args['std_mark'] : '';` 
$std_mark is always empty cause $this->options have already been set  at function _set_default_options() and isset($this->options[$field['id']]) is always true.

[This code](https://github.com/signo/Redux-Framework/blob/3bfed84c1c97c5679b7ae10909e06be059b80c55/options/defaults.php#L836-L837) always sets checkbox value to std when checkbox uncheked and std_show == true.
We had this [issue](https://github.com/ghost1227/Redux-Framework/issues/43#issuecomment-12156692)
It happens because [your file](https://github.com/signo/Redux-Framework/blob/3bfed84c1c97c5679b7ae10909e06be059b80c55/options/defaults.php#L492) missmatches current [master defaults.php](%28https://github.com/ghost1227/Redux-Framework/blob/master/options/defaults.php#L499%29) (somehow it  isn't shown in request file changes)
If checkbox is unncheked it isn't sent in the $_POST and next time page load it would be [rewrote with std](https://github.com/signo/Redux-Framework/blob/3bfed84c1c97c5679b7ae10909e06be059b80c55/options/defaults.php#L836-L837)

Commit:
Restored test for for null for $this->sections that somehow got deleted in an update.

signo commented:
The problem you are describing won't happen since all the changes made to _set_default_options() where reverted in this pull request.

Commit:
get() now returns the std value if std_show = on
added _get_std() to centralize getting the std value in one place
added a cache for the std values, preserving the old _default_values() functionality.
made code more robust

Commit:
fixed bug where get method was called with 'std' value instead of 'id' in function _field_input

evertiro commented:
This pull has gotten severely convoluted. Can you please resubmit based on a clean pull?
